skfda/representation/basis/_custom_basis.py

```diff
@@ -0,0 +1,219 @@
+        # whose codomain is not 1-dimensional and it can be done because
+        # checking linear independence in (R^n)^k is equivalent to doing
+        # it in R^(nk).
+        coord_matrix = fdata.data_matrix.reshape(
+            fdata.data_matrix.shape[0],
+            -1,
+        )
+        return self._check_linearly_independent_matrix(coord_matrix)
+
+    @_check_linearly_independent.register
+    def _check_linearly_independent_basis(self, fdata: FDataBasis) -> None:
+        return self._check_linearly_independent_matrix(fdata.coefficients)
+
+    def _check_linearly_independent_matrix(self, matrix: NDArrayFloat) -> None:
+        """Check if the functions are linearly independent."""
+        if matrix.shape[0] > matrix.shape[1]:
+            raise ValueError(
+                "There are more functions than the maximum dimension of the "
+                "space that they could generate.",
+            )
+
+        rank = np.linalg.matrix_rank(matrix)
+        if rank != matrix.shape[0]:
+            raise ValueError(
+                "There are only {rank} linearly independent "
+                "functions".format(
+                    rank=rank,
+                ),
+            )
+
+    def _derivative_basis_and_coefs(
+        self: T,
+        coefs: NDArrayFloat,
+        order: int = 1,
+    ) -> Tuple[T, NDArrayFloat]:
+        deriv_fdata = self.fdata.derivative(order=order)
+
+        return self._create_subspace_basis_coef(deriv_fdata, coefs)
+
+    @multimethod.multidispatch
+    def _create_subspace_basis_coef(
+        self: T,
+        fdata: FData,
+        coefs: np.ndarray,
+    ) -> Tuple[T, NDArrayFloat]:
+        """
+        Create a basis of the subspace generated by the given functions.
+
+        Args:
+            fdata: The resulting basis will span the subspace generated
+                by these functions.
+            coefs: Coefficients of some functions in the given fdata.
+                These coefficients will be transformed into the coefficients
+                of the same functions in the resulting basis.
+        """
+        raise ValueError(
+            "Unexpected type of functional data object: {type}.".format(
+                type=type(fdata),
+            ),
+        )
+
+    @_create_subspace_basis_coef.register
+    def _create_subspace_basis_coef_grid(
+        self: T,
+        fdata: FDataGrid,
+        coefs: np.ndarray,
+    ) -> Tuple[T, NDArrayFloat]:
+
+        # Reshape to a bidimensional matrix. This can be done because
+        # working in (R^n)^k is equivalent to working in R^(nk) when
+        # it comes to linear independence and basis.
+        data_matrix_reshaped = fdata.data_matrix.reshape(
+            fdata.data_matrix.shape[0],
+            -1,
+        )
+        # If the basis formed by the derivatives has maximum rank,
+        # we can just return that
+        rank = np.linalg.matrix_rank(data_matrix_reshaped)
```

Comment:
Couldn't you use `_check_linearly dependent` and thus share code for basis and grid?